### PR TITLE
Mutable iterators

### LINKED ISSUE #69: RingBufferMutIterator doesn't actually impl Iterator
Currently, `RingBufferMutIterator` doesn't actually `impl Iterator` and that's a bit annoying when you want to do something like `for element in buffer.iter_mut() {}`. This was apparently removed in eb7ff21e684bc5309e6f65d8be815df3853d02cc but it's now just really misleading as the name implies that it's an `Iterator` but it's actually not. Can this again be made to `impl Iterator`?

### PR DESCRIPTION
**Pull request by jonay2000**
Closes #69

### CODE REVIEW

**Comment by jonay2000** [comment]
image

and that's why you test

**Comment by jonay2000** [comment]
Would close #69 if it worked, but this implementation (and any implementation I can make with my knowledge) doesn't pass the miri tests.

**Comment by jonay2000** [comment]
@svenstaro

**Comment by jonay2000** [comment]
I have gained some knowledge! And I've added a lot of unsafe code that uses *mut instead of &mut (the latter is detected as undefined behavior by miri, and I think rightfully so). However, with the last commit miri is happy!

File: src/lib.rs [hunk 1]
```diff
@@ -314,22 +317,41 @@ mod tests {
...
+
+            assert_eq!(vec![3, 4], b.to_vec())
+        }
+
+        run_test_iter_mut_wrap(AllocRingBuffer::with_capacity(2));
+        run_test_iter_mut_wrap(ConstGenericRingBuffer::<i32, 2>::new());
+    }
+
+    #[test]
+    fn test_iter_mut_miri_fail() {
```
Comment (jonay2000) [inline src/lib.rs:338]: This is the most important function. Without all the unsafe functions (like `ptr_len` and `ptr_capacity` and `ptr_get_mut` which *only* use raw pointers, no references anywhere!!! this is safe. However, as soon as we construc a mutable reference anywhere, miri detects it as aliasing (even though we're technically not able to use both at the same time). This is similar to what `std::vec::Vec` does, however it uses raw pointers everywhere anyway. I think all the unsafe code is very thoroughly tested, and miri passes so I'm quite confident about its correctness.

**Review by NULLx76 — APPROVED** [review]
Fixed some things in #74

**Pull request merged**